crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,477 @@
+};
+use petgraph::{
+    graph::NodeIndex,
+    prelude::StableDiGraph,
+};
+
+use crate::{
+    collision_manager::{
+        basic::BasicCollisionManagerStorage,
+        CollisionReason,
+    },
+    error::Error,
+    ports::TxPoolPersistentStorage,
+    selection_algorithms::ratio_tip_gas::RatioTipGasSelectionAlgorithmStorage,
+};
+
+use super::{
+    RemovedTransactions,
+    Storage,
+    StorageData,
+};
+
+pub struct GraphStorage {
+    /// The configuration of the graph
+    config: GraphConfig,
+    /// The graph of transactions
+    graph: StableDiGraph<StorageData, ()>,
+    /// Coins -> Transaction that currently create the UTXO
+    coins_creators: HashMap<UtxoId, NodeIndex>,
+    /// Contract -> Transaction that currently create the contract
+    contracts_creators: HashMap<ContractId, NodeIndex>,
+}
+
+pub struct GraphConfig {
+    /// The maximum number of transactions per dependency chain
+    pub max_dependent_txn_count: u64,
+}
+
+impl GraphStorage {
+    /// Create a new graph storage
+    pub fn new(config: GraphConfig) -> Self {
+        Self {
+            config,
+            graph: StableDiGraph::new(),
+            coins_creators: HashMap::new(),
+            contracts_creators: HashMap::new(),
+        }
+    }
+}
+
+impl GraphStorage {
+    /// Remove a node and all its dependent sub-graph.
+    /// Edit the data of dependencies transactions accordingly.
+    /// Returns the removed transactions.
+    fn remove_node_and_dependent_sub_graph(
+        &mut self,
+        root_id: NodeIndex,
+    ) -> Result<Vec<PoolTransaction>, Error> {
+        let Some(root) = self.graph.node_weight(root_id) else {
+            return Ok(vec![])
+        };
+        let gas_removed = root.dependents_cumulative_gas;
+        let tip_removed = root.dependents_cumulative_tip;
+        self.reduce_dependencies_cumulative_gas_and_tip(
+            root_id,
+            gas_removed,
+            tip_removed,
+        )?;
+        self.remove_dependent_sub_graph(root_id)
+    }
+    fn reduce_dependencies_cumulative_gas_and_tip(
+        &mut self,
+        root_id: NodeIndex,
+        gas_reduction: u64,
+        tip_reduction: u64,
+    ) -> Result<(), Error> {
+        let Some(root) = self.graph.node_weight_mut(root_id) else {
+            return Err(Error::Storage(format!(
+                "Node with id {:?} not found",
+                root_id
+            )));
+        };
+        root.dependents_cumulative_gas =
+            root.dependents_cumulative_gas.saturating_sub(gas_reduction);
+        root.dependents_cumulative_tip =
+            root.dependents_cumulative_tip.saturating_sub(tip_reduction);
+        for dependency in self.get_dependencies(root_id)? {
+            self.reduce_dependencies_cumulative_gas_and_tip(
+                dependency,
+                gas_reduction,
+                tip_reduction,
+            )?;
+        }
+        Ok(())
+    }
+    fn remove_dependent_sub_graph(
```

Comment:
It looks like we don't call `clear_cache`, so I assume we have a lot of trash in the memory after this operation

Reply:
Right, Done. We should really test all the caches in the tests and even better link the caches to the storage with some kind of hook to be sure they are synced. Open an issue on that